File: backend/app/schemas/portfolio.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class ConnectExchangeRequest(BaseModel):
    exchange: str = Field(..., pattern=r"^(binance|bybit|kucoin)$")
    api_key: str = Field(..., min_length=10)
    api_secret: str = Field(..., min_length=10)
    api_passphrase: str | None = Field(default=None)
    label: str = Field(default="My Account", max_length=100)

    @property
    def passphrase(self) -> str | None:
        return self.api_passphrase


class ExchangeAccountResponse(BaseModel):
    id: UUID
    exchange: str
    label: str
    is_active: bool
    last_synced_at: datetime | None
    sync_error: str | None

    model_config = {"from_attributes": True}


class HoldingItem(BaseModel):
    symbol: str
    name: str | None
    quantity: float
    avg_buy_price: float
    current_price: float
    value_usd: float
    unrealized_pnl_usd: float
    unrealized_pnl_pct: float
    allocation_pct: float


class PortfolioSummary(BaseModel):
    total_value_usd: float
    total_pnl_usd: float
    total_pnl_pct: float
    daily_change_usd: float
    daily_change_pct: float
    holdings: list[HoldingItem]
    last_updated: datetime


class TradeResponse(BaseModel):
    id: UUID
    symbol: str
    side: str
    price: float
    quantity: float
    quote_quantity: float
    fee: float
    fee_asset: str | None
    realized_pnl: float | None
    executed_at: datetime

    model_config = {"from_attributes": True}


class SyncStatusResponse(BaseModel):
    exchange_account_id: str
    status: str
    trades_synced: int
    message: str

File: lib/domain/entities/portfolio.dart
class HoldingEntity {
  final String symbol;
  final String? name;
  final double quantity;
  final double avgBuyPrice;
  final double currentPrice;
  final double valueUsd;
  final double unrealizedPnlUsd;
  final double unrealizedPnlPct;
  final double allocationPct;

  const HoldingEntity({
    required this.symbol,
    this.name,
    required this.quantity,
    required this.avgBuyPrice,
    required this.currentPrice,
    required this.valueUsd,
    required this.unrealizedPnlUsd,
    required this.unrealizedPnlPct,
    required this.allocationPct,
  });

  bool get isProfit => unrealizedPnlUsd >= 0;
}

class PortfolioSummaryEntity {
  final double totalValueUsd;
  final double totalPnlUsd;
  final double totalPnlPct;
  final double dailyChangeUsd;
  final double dailyChangePct;
  final List<HoldingEntity> holdings;
  final DateTime lastUpdated;

  const PortfolioSummaryEntity({
    required this.totalValueUsd,
    required this.totalPnlUsd,
    required this.totalPnlPct,
    required this.dailyChangeUsd,
    required this.dailyChangePct,
    required this.holdings,
    required this.lastUpdated,
  });

  bool get isPositive => totalPnlUsd >= 0;
  bool get isDailyPositive => dailyChangeUsd >= 0;
}

class ExchangeAccountEntity {
  final String id;
  final String exchange;
  final String label;
  final bool isActive;
  final DateTime? lastSyncedAt;
  final String? syncError;

  const ExchangeAccountEntity({
    required this.id,
    required this.exchange,
    required this.label,
    required this.isActive,
    this.lastSyncedAt,
    this.syncError,
  });
}

class TradeEntity {
  final String id;
  final String symbol;
  final String side;
  final double price;
  final double quantity;
  final double quoteQuantity;
  final double fee;
  final double? realizedPnl;
  final DateTime executedAt;

  const TradeEntity({
    required this.id,
    required this.symbol,
    required this.side,
    required this.price,
    required this.quantity,
    required this.quoteQuantity,
    required this.fee,
    this.realizedPnl,
    required this.executedAt,
  });

  bool get isBuy => side == 'buy';
  bool get isProfit => (realizedPnl ?? 0) > 0;
}
